components/raftstore/src/store/fsm/peer.rs

```diff
@@ -2053,25 +2062,46 @@ where
             false
         } else {
             // Destroy the peer fsm directly
-            self.destroy_peer(false);
-            true
+            self.destroy_peer(false)
         }
     }
 
-    fn destroy_peer(&mut self, merged_by_target: bool) {
+    fn destroy_peer(&mut self, merged_by_target: bool) -> bool {
         fail_point!("destroy_peer");
+        // Mark itself as pending_remove
+        self.fsm.peer.pending_remove = true;
+
+        if self.fsm.peer.has_unpersisted_ready() {
```

Comment:
It seems possible `destroy_peer` is called twice. For example, it's suggested to destroy itself by gc message and raft entry conf remove apply.